File: Creational/Builder/Builder.ts
interface Builder {
  buildPartA(): void;
  buildPartB(): void;
  buildPartC(): void;
}

class MainProduct {
  public parts: string[] = [];

  public listParts(): void {
    console.log(`Product parts: ${this.parts.join(', ')}\n`);
  }
}

class BuilderA implements Builder {
  private product: MainProduct;

  constructor() {
    this.reset();
  }

  public reset(): void {
    this.product = new MainProduct();
  }

  public buildPartA(): void {
    this.product.parts.push('PartA1');
  }

  public buildPartB(): void {
    this.product.parts.push('PartB1');
  }

  public buildPartC(): void {
    this.product.parts.push('PartC1');
  }

  public buildProduct(): MainProduct {
    const result = this.product;
    this.reset();
    return result;
  }
}

class Director {
  private builder: Builder;

  public setBuilder(builder: Builder): void {
    this.builder = builder;
  }

  public buildProduct(): void {
    this.builder.buildPartA();
    this.builder.buildPartB();
    this.builder.buildPartB();
  }
}

function main() {
  const director = new Director();
  const builder = new BuilderA();
  director.setBuilder(builder);
}

File: Builder/Builder.py
from __future__ import annotations
from abc import ABCMeta, abstractmethod


class Builder(metaclass=ABCMeta):
    @abstractmethod
    def build_part_A(self) -> None:
        pass

    @abstractmethod
    def build_part_B(self) -> None:
        pass

    @abstractmethod
    def build_part_C(self) -> None:
        pass


class Product:
    def __init__(self) -> None:
        self.parts = []

    def list_parts(self) -> None:
        print(f"Product parts: {', '.join(self.parts)}", end="")


class BuilderA(Builder):
    def __init__(self) -> None:
        self.reset()

    def reset(self) -> None:
        self._product = Product1()

    def build_part_a(self) -> None:
        self._product.add("PartA1")

    def build_part_b(self) -> None:
        self._product.add("PartB1")

    def build_part_c(self) -> None:
        self._product.add("PartC1")

    @property
    def build_product(self) -> Product1:
        product = self._product
        self.reset()
        return product


class Director:
    def __init__(self) -> None:
        self._builder = None

    @property
    def builder(self) -> Builder:
        return self._builder

    @builder.setter
    def builder(self, builder: Builder) -> None:
        self._builder = builder

    def build_product(self) -> None:
        self._builder.build_part_A()
        self._builder.build_part_B()
        self._builder.build_part_C()


def main():
    director = Director()
    builder = BuilderA()
    director.builder = builder

if __name__="__main__":
    main()
